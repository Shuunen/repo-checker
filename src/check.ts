/* c8 ignore next */
import { ellipsis, green, red, yellow } from 'shuutils'
import type { ProjectData } from './constants'
import { DependencyCruiserFile, EditorConfigFile, EsLintFile, GitFile, GithubWorkflowFile, LicenseFile, NpmRcFile, NvmrcFile, NycRcFile, PackageJsonFile, ReadmeFile, RenovateFile, RepoCheckerConfigFile, TailwindFile, TravisFile, TsConfigFile } from './files/index.js'
import { log } from './logger'
import { augmentData, getProjectFolders } from './utils'

interface Indicators { failed: string[]; passed: string[]; warnings: string[] }

const checkers = [
  DependencyCruiserFile,
  EditorConfigFile,
  EsLintFile,
  GitFile,
  GithubWorkflowFile,
  LicenseFile,
  NpmRcFile,
  NvmrcFile,
  NycRcFile,
  PackageJsonFile,
  ReadmeFile,
  RenovateFile,
  RepoCheckerConfigFile,
  TailwindFile,
  TravisFile,
  TsConfigFile,
]

function reportLog (color: (string: string) => string, count: number, message: string): void {
  const line = `â€£ ${count} check${count > 1 ? 's' : ''} ${message}`
  /* c8 ignore next */
  log.info(count === 0 ? line : color(line))
}

interface CheckOptions {
  canFailStop?: boolean
  canFix?: boolean
  canForce?: boolean
  canThrow?: boolean
  data: ProjectData
  folderPath: string
}

function report ({ failed = [], passed = [], warnings = [] }: Indicators): void {
  log.info('Report :')
  reportLog(green, passed.length, 'are successful')
  reportLog(yellow, warnings.length, 'triggered warnings')
  reportLog(red, failed.length, 'are problematic')
  /* c8 ignore next */
}

// eslint-disable-next-line max-statements
export async function check ({ canFailStop = false, canFix = false, canForce = false, canThrow = true, data, folderPath }: CheckOptions): Promise<Indicators> {
  const folders = await getProjectFolders(folderPath)
  let passed: string[] = []
  let warnings: string[] = []
  let failed: string[] = []
  log.options.isActive = !data.isQuiet
  /* c8 ignore next */
  if (folders.length === 0) log.warn('no folder to check', folderPath)
  /* eslint-disable no-await-in-loop */
  for (const folder of folders) {
    if (canFailStop && failed.length > 0) {
      log.warn('stop checking other folders because of failures & --fail-stop option')
      break
    }
    log.info('Checking folder :', folder)
    const dataFolder = await augmentData(folder, data, folders.length > 1)
    for (const Checker of checkers) { // eslint-disable-line @typescript-eslint/naming-convention
      const instance = new Checker(folder, dataFolder, canFix, canForce)
      await instance.start()
      await instance.end()
      passed = [...passed, ...instance.passed]
      warnings = [...warnings, ...instance.warnings]
      failed = [...failed, ...instance.failed]
    }
  }
  /* eslint-enable no-await-in-loop */
  report({ failed, passed, warnings })
  const maxLogLength = 100
  if (canThrow && failed.length > 0) throw new Error(`failed at validating at least one rule in one folder : ${ellipsis(failed.join(', '), maxLogLength)}`)
  return { failed, passed, warnings }
}
