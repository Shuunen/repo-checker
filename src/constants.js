import { join } from 'path'

export const home = process.env.HOME
export const dataFileName = 'repo-checker.config.js'
export const repoCheckerPath = join(__dirname, '..')
export const templatePath = join(repoCheckerPath, 'templates')
export const dataFileHomePath = join(home, dataFileName)

export const dataDefaults = {
  auto_merge: true,
  ban_sass: true,
  dev_deps_only: true,
  license: 'GPL-3.0',
  max_size_ko: 10,
  user_mail: 'romain.racamier@gmail.com',
  user_name: 'Romain Racamier-Lafon',
}
