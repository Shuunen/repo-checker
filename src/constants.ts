import { join } from 'path'

/* istanbul ignore next */
export const home = process.env.HOME ?? '' // when does HOME is not defined ?
export const dataFileName = 'repo-checker.config.js'
/* istanbul ignore next */
export const repoCheckerPath = process.env.pwd ?? process.cwd()
export const templatePath = join(repoCheckerPath, 'templates')
export const dataFileHomePath = join(home, dataFileName)

export class ProjectData {
  auto_merge = true
  ban_sass = true
  dev_deps_only = true
  is_module = false
  license = 'GPL-3.0'
  max_size_ko = 1
  npm_package = false
  package_name = ''
  quiet = false
  repo_id = 'a-great-repo'
  use_typescript = false
  use_vue = false
  user_id = 'Shuunen'
  user_id_lowercase = 'shuunen'
  user_mail = 'romain.racamier@gmail.com'
  user_name = 'Romain Racamier-Lafon'
  web_published = false
  web_url = 'https://my-website.com'

  constructor (data: Partial<ProjectData> = {}) {
    Object.assign(this, data)
  }
}

export const dataDefaults = new ProjectData()
