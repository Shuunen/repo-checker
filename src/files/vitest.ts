import { FileBase } from '../file.ts'

// eslint-disable-next-line no-export
export class VitestFile extends FileBase {
  // eslint-disable-next-line jsdoc/require-jsdoc
  // eslint-disable-next-line max-lines-per-function
  public async start() {
    if (!this.data.isUsingVitest) return
    const hasMtsFile = await this.fileExists('vitest.config.mts')
    const hasJsFile = await this.fileExists('vitest.config.js')
    const hasTsFile = await this.fileExists('vitest.config.ts')
    const hasMjsFile = await this.fileExists('vitest.config.mjs')
    if (hasMtsFile) await this.inspectFile('vitest.config.mts')
    /* c8 ignore next 3 */
    if (hasJsFile) await this.inspectFile('vitest.config.js')
    if (hasTsFile) await this.inspectFile('vitest.config.ts')
    if (hasMjsFile) await this.inspectFile('vitest.config.mjs')
    this.couldContains('a pool thread setting', /pool: 'threads'/u)
  }
}
