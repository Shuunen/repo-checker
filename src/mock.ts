/* eslint-disable jsdoc/require-jsdoc */
import path from 'node:path'
import { clone, sleep } from 'shuutils'
import type { Indicators } from './check.ts'
import type { FileBase } from './file.ts'
import { join } from './utils.ts'

const fileBaseKeysToDelete: (keyof FileBase)[] = ['fileExists', 'checkFileExists', 'folderPath', 'inspectFile', 'updateFile', 'initFile']
const indicatorsKeys: ReadonlyArray<keyof Indicators> = ['failed', 'passed', 'warnings'] as const

function cleanStringForSnap(input: string) {
  return (
    input
      // replace repo-check-1-40 by repo-check-x-yy
      .replace(/repo-check-\d-\d+/gu, 'repo-check-x-yy')
      // replace package-json-main-file-size-63ko by package-json-main-file-size-XYZko
      .replace(/package-json-main-file-size-\d+ko/gu, 'package-json-main-file-size-XYZko')
  )
}

// eslint-disable-next-line no-restricted-syntax
export function cleanUnknownValueForSnap<Type>(input: Type): Type {
  /* c8 ignore next 8 */
  /* eslint-disable @typescript-eslint/consistent-type-assertions, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-type-assertion */
  if (typeof input === 'string') return cleanStringForSnap(input) as Type
  if (Array.isArray(input)) return input.map(item => cleanUnknownValueForSnap(item)) as Type
  if (typeof input === 'object') return JSON.parse(cleanStringForSnap(JSON.stringify(input))) as Type
  // eslint-disable-next-line no-restricted-syntax
  throw new Error(`cleanInstanceValueForSnap: unknown type ${typeof input}`)
  /* eslint-enable @typescript-eslint/consistent-type-assertions, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-type-assertion */
}

export async function promiseValue<Type>(value: Type) {
  await sleep(1)
  return value
}

export async function promiseTrue() {
  await sleep(1)
  return true
}

export async function promiseFalse() {
  await sleep(1)
  return false
}

export async function promiseVoid() {
  await sleep(1)
}

// eslint-disable-next-line unicorn/prefer-module
export const sourceFolder = path.dirname(__filename)

export const vueProjectFolder = join(sourceFolder, 'mocks', 'vueProject')

export const tsProjectFolder = join(sourceFolder, 'mocks', 'tsProject')

export const mocksProjectsFolder = join(sourceFolder, 'mocks')

export function cleanIndicatorsForSnap(record?: Readonly<Record<string, unknown>>) {
  if (!record) return record
  const clean = clone(record)
  for (const key of indicatorsKeys) if (key in clean) clean[key] = cleanUnknownValueForSnap(clean[key])
  return clean
}

// eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
export function cleanInstanceForSnap(instance: Readonly<FileBase>, ...bonusKeys: readonly string[]) {
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  const clean = clone<Partial<FileBase>>(instance) as Record<string, unknown>
  const keys = [...fileBaseKeysToDelete, ...bonusKeys] // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
  for (const key of keys) if (key in clean) delete clean[key]
  return cleanIndicatorsForSnap(clean)
}
