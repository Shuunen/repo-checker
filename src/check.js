import * as files from './files'
import { log } from './logger'
import { augmentData, getGitFolders } from './utils'

const Checkers = Object.keys(files).map(f => files[f])

function report (nbPassed, nbFailed) {
  log.info('Report :')
  log.setIndentLevel(1)
  log.test(nbPassed > 0, `${nbPassed} test(s) passed successfully`, false, true)
  log.test(nbFailed === 0, `${nbFailed} test(s) failed`, false, true)
  log.line()
  log.setIndentLevel(0)
  if (nbFailed > 0) throw new Error('failed at validating at least one rule in one folder')
}

export async function check (folderPath, data, doFix, doForce) {
  const folders = await getGitFolders(folderPath)
  let nbPassed = 0
  let nbFailed = 0
  for (const folder of folders) {
    log.info('Checking folder :', folder)
    log.setIndentLevel(1)
    const dataFolder = await augmentData(folder, data)
    for (const Checker of Checkers) {
      const instance = new Checker(folder, dataFolder, doFix, doForce)
      await instance.start()
      await instance.end()
      nbPassed += instance.nbPassed
      nbFailed += instance.nbFailed
    }
    log.line()
    log.setIndentLevel(0)
  }
  report(nbPassed, nbFailed)
}
