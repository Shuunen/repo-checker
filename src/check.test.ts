import { expect, it } from 'vitest'
import { check } from './check'
import { ProjectData, repoCheckerPath } from './constants'
import { cleanIndicatorsForSnap, cleanUnknownValueForSnap, mocksProjectsFolder, tsProjectFolder } from './mock'

it('check A repo-checker folder fails with low max size', async () => {
  const data = new ProjectData({ isPublishedPackage: true, isQuiet: true, maxSizeKo: 2 })
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions, @typescript-eslint/no-unsafe-type-assertion
  const message = await check({ data, folderPath: repoCheckerPath }).catch((error: unknown) => (error as Error).message)
  expect(cleanUnknownValueForSnap(message)).toMatchSnapshot()
})

it('check B repo-checker folder succeed', async () => {
  const data = new ProjectData({ isModule: true, isPublishedPackage: true, isQuiet: true, maxSizeKo: 120 })
  const indicators = await check({ data, folderPath: repoCheckerPath })
  expect(cleanIndicatorsForSnap(indicators)).toMatchSnapshot()
})

it('check C ts-project folder', async () => {
  const data = new ProjectData({ isQuiet: true })
  const indicators = await check({ canThrow: false, data, folderPath: tsProjectFolder })
  expect(cleanIndicatorsForSnap(indicators)).toMatchSnapshot()
})

it('check D mocks-projects folders and throw', async () => {
  const data = new ProjectData({ isQuiet: true })
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions, @typescript-eslint/no-unsafe-type-assertion
  const message = await check({ canFailStop: true, data, folderPath: mocksProjectsFolder }).catch((error: unknown) => (error as Error).message)
  expect(message).toMatchSnapshot()
})

it('check E mocks-projects folders and not throw', async () => {
  const data = new ProjectData({ isQuiet: true })
  const indicators = await check({ canFailStop: true, canThrow: false, data, folderPath: mocksProjectsFolder })
  expect(cleanIndicatorsForSnap(indicators)).toMatchSnapshot()
})
