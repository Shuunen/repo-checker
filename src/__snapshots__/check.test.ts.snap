// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`check A folder fails with low max size 1`] = `"failed at validating at least one rule in one folder : package-json-main-file-size-70ko-should-be-less-or-equal-to-max-size-allowed-2ko"`;

exports[`check B folder succeed 1`] = `
{
  "failed": [],
  "passed": [
    "use-dependency-cruiser",
    "has-a-dependency-cruiser-config-file",
    "has-a-editorconfig-file",
    "editorconfig-has-space-indent",
    "editorconfig-has-indent-size-of-2",
    "editorconfig-has-unix-style-line-endings",
    "editorconfig-has-utf-8-encoding",
    "editorconfig-has-whitespace-trailing",
    "editorconfig-has-final-new-line-rule",
    "editorconfig-has-no-specific-html-indent-rule",
    "has-no-xo-config-js-file",
    "has-a-eslint-config-cjs-file",
    "avoid-main-branch-reference-use-master-instead-git-bclean",
    "has-a-gitignore-file",
    "gitignore-has-node-modules",
    "gitignore-has-no-pnpm-lock-exclusion",
    "github-workflows-ci-yml-has-a-checkout-step-in-ci-workflow",
    "github-workflows-ci-yml-has-a-recent-checkout-version",
    "github-workflows-ci-yml-has-a-node-step-in-ci-workflow",
    "github-workflows-ci-yml-has-a-pnpm-setup-step",
    "github-workflows-ci-yml-has-a-recent-pnpm-version-8-or-9",
    "github-workflows-ci-yml-has-no-frozen-lockfile-flag",
    "github-workflows-ci-yml-has-no-main-branch-reference",
    "github-workflows-ci-yml-has-a-recent-node-version",
    "has-a-license-file",
    "license-has-a-gpl-title",
    "license-has-a-version-3-subtitle",
    "has-a-npmrc-file",
    "npmrc-has-enable-pre-post-scripts-option",
    "has-a-nvmrc-file",
    "nvmrc-has-a-recent-lts-node-version",
    "has-a-package-json-file",
    "package-json-main-file-maximum-size-is-specified-in-data-file-ex-max-size-ko-100",
    "package-json-has-a-dist-repo-check-min-cjs-file",
    "package-json-main-file-specified-in-package-json-dist-repo-check-min-cjs-exists-on-disk-be-sure-to-build-before-run-repo-check",
    "package-json-main-file-size-70ko-should-be-less-or-equal-to-max-size-allowed-120ko",
    "package-json-has-a-schema-declaration",
    "package-json-has-a-bugs-property",
    "package-json-has-a-description-property",
    "package-json-has-a-files-property",
    "package-json-has-a-homepage-property",
    "package-json-has-a-keywords-property",
    "package-json-has-a-private-property",
    "package-json-has-a-repository-property",
    "package-json-has-a-author-property",
    "package-json-has-a-name-property",
    "package-json-has-a-version-property",
    "package-json-has-a-license-property",
    "package-json-has-a-gpl-3-0-license",
    "package-json-has-a-typescript-build-or-check",
    "package-json-has-a-script-section",
    "package-json-has-a-pre-script-for-version-automation",
    "package-json-has-a-post-script-for-version-automation",
    "package-json-has-watchlist-eager-param",
    "package-json-has-a-depcruise-usage",
    "package-json-has-a-unique-mark-task",
    "package-json-has-pnpm-instead-of-npm-run",
    "package-json-has-a-check-script",
    "package-json-has-no-ci-script",
    "package-json-has-a-final-echo-for-task-build",
    "package-json-has-a-final-echo-for-task-check",
    "package-json-has-a-final-echo-for-task-lint",
    "package-json-has-a-final-echo-for-task-test",
    "package-json-has-no-sass-dependency-fat-useless",
    "package-json-has-no-cross-var-dependency-old-deprecated",
    "package-json-has-no-tslint-dependency-deprecated",
    "package-json-has-no-eslint-plugin-promise-5-dependency-require-eslint-7",
    "package-json-has-no-patch-precision",
    "package-json-one-unit-testing-dependency-from-vitest-mocha-uvu",
    "package-json-one-coverage-dependency-from-nyc-c8-v8",
    "package-json-has-an-eslint-dependency",
    "package-json-has-no-eslint8-dependency",
    "package-json-has-eslint-cache-flag",
    "package-json-has-no-eslint-ignore-flag-solution-1-just-remove-it-useless-most-of-the-time-check-debugeslint-cli-engine-npx-eslint-to-see-linted-files-or-solution-2-use-ignore-patterns-inside-eslintrc-json-the-objective-here-is-to-let-the-eslint-cli-vscode-eslint-use-the-same-config",
    "package-json-has-no-fat-color-dependency-use-shuutils-or-nanocolors",
    "package-json-has-no-fat-fs-extra-dependency-use-native-fs",
    "package-json-has-no-utopian-shuunen-stack-dependency",
    "package-json-has-no-fat-slow-jsdom-dependency-use-happy-dom-instead",
    "package-json-has-no-fat-task-runner-use-pnpm-xyz-pnpm-abc-for-sequential-or-zero-deps-package-npm-parallel",
    "readme-md-has-a-title",
    "readme-md-has-a-logo-image",
    "readme-md-has-no-link-to-deprecated-netlify-com",
    "readme-md-has-no-links-without-https-scheme",
    "readme-md-has-no-crlf-windows-carriage-return",
    "readme-md-has-no-star-flavored-list",
    "readme-md-has-a-project-license-badge",
    "readme-md-has-a-code-climate-maintainability-badge",
    "readme-md-has-a-npm-monthly-downloads-badge",
    "readme-md-has-a-npm-version-badge",
    "readme-md-has-a-publish-size-badge",
    "readme-md-has-a-install-size-badge",
    "readme-md-has-no-deprecated-bettercodehub-com-badge",
    "readme-md-has-no-deprecated-david-dm-org-badge",
    "readme-md-has-no-deprecated-lgtm-com-badge",
    "readme-md-has-no-deprecated-scrutinizer-ci-com-badge",
    "readme-md-has-no-deprecated-libraries-io-badge",
    "readme-md-has-a-thanks-section",
    "readme-md-has-a-thanks-to-shields-io",
    "readme-md-has-no-remaining-thanks-to-travis-ci-com",
    "readme-md-has-a-thanks-to-github",
    "readme-md-has-no-remaining-thanks-to-netlify",
    "readme-md-has-a-thanks-to-arg",
    "readme-md-has-no-remaining-thanks-to-ava",
    "readme-md-has-no-remaining-thanks-to-c8",
    "readme-md-has-no-remaining-thanks-to-chokidar",
    "readme-md-has-no-remaining-thanks-to-cypress-io",
    "readme-md-has-a-thanks-to-dependency-cruiser",
    "readme-md-has-a-thanks-to-esbuild",
    "readme-md-has-a-thanks-to-eslint",
    "readme-md-has-no-remaining-thanks-to-mocha",
    "readme-md-has-no-remaining-thanks-to-npm-run-all",
    "readme-md-has-no-remaining-thanks-to-npm-parallel",
    "readme-md-has-no-remaining-thanks-to-nuxt",
    "readme-md-has-no-remaining-thanks-to-nyc",
    "readme-md-has-no-remaining-thanks-to-reef",
    "readme-md-has-a-thanks-to-repo-checker",
    "readme-md-has-no-remaining-thanks-to-rollup",
    "readme-md-has-no-remaining-thanks-to-servor",
    "readme-md-has-no-remaining-thanks-to-showdown",
    "readme-md-has-a-thanks-to-shuutils",
    "readme-md-has-no-remaining-thanks-to-tailwind-css",
    "readme-md-has-no-remaining-thanks-to-tsup",
    "readme-md-has-no-remaining-thanks-to-uv-u",
    "readme-md-has-a-thanks-to-v8",
    "readme-md-has-a-thanks-to-vitest",
    "readme-md-has-no-remaining-thanks-to-vite",
    "readme-md-has-no-remaining-thanks-to-vue",
    "readme-md-has-a-thanks-to-watchlist",
    "readme-md-has-no-remaining-thanks-to-xo",
    "readme-md-has-a-star-chart",
    "readme-md-has-a-page-views-counter",
    "has-a-renovate-json-file",
    "renovate-json-has-a-schema-declaration",
    "renovate-json-has-an-extends-section",
    "renovate-json-has-a-base-config",
    "renovate-json-has-a-dashboard-setting-to-false",
    "renovate-json-has-an-auto-merge-preset",
    "renovate-json-has-a-preserve-semver-ranges-preset",
    "has-a-repo-checker-json-file",
    "tsconfig-json-has-a-schema-declaration",
    "tsconfig-json-has-an-include-section",
    "tsconfig-json-has-a-allow-unreachable-code-compiler-option",
    "tsconfig-json-has-a-allow-unused-labels-compiler-option",
    "tsconfig-json-has-a-check-js-compiler-option",
    "tsconfig-json-has-a-es-module-interop-compiler-option",
    "tsconfig-json-has-a-exact-optional-property-types-compiler-option",
    "tsconfig-json-has-a-force-consistent-casing-in-file-names-compiler-option",
    "tsconfig-json-has-a-no-fallthrough-cases-in-switch-compiler-option",
    "tsconfig-json-has-a-no-implicit-any-compiler-option",
    "tsconfig-json-has-a-no-implicit-override-compiler-option",
    "tsconfig-json-has-a-no-implicit-returns-compiler-option",
    "tsconfig-json-has-a-no-property-access-from-index-signature-compiler-option",
    "tsconfig-json-has-a-no-unchecked-indexed-access-compiler-option",
    "tsconfig-json-has-a-no-unused-locals-compiler-option",
    "tsconfig-json-has-a-no-unused-parameters-compiler-option",
    "tsconfig-json-has-a-skip-lib-check-compiler-option",
    "tsconfig-json-has-a-strict-compiler-option",
    "tsconfig-json-has-a-verbatim-module-syntax-compiler-option",
    "tsconfig-json-has-a-out-dir-compiler-option",
    "tsconfig-json-has-a-module-resolution-compiler-option",
    "tsconfig-json-has-a-non-empty-lib-compiler-option",
    "tsconfig-json-my-folder-is-not-needed-in-include-section-my-folder-is-enough",
  ],
  "warnings": [
    "readme-md-could-have-a-demo-image-or-gif-demofolder-any-file-ext",
  ],
}
`;

exports[`check C data/tsProject 1`] = `
{
  "failed": [
    "has-a-editorconfig-file",
    "has-a-gitignore-file",
    "has-a-license-file",
    "has-a-nvmrc-file",
    "package-json-does-not-have-a-name-property-name",
    "package-json-does-not-have-a-typescript-build-or-check-btscb",
    "package-json-does-not-have-prepare-instead-of-prepublish-deprecated-prepublish",
  ],
  "passed": [
    "avoid-main-branch-reference-use-master-instead-git-bclean",
    "has-a-package-json-file",
    "package-json-has-a-description-property",
    "package-json-has-a-author-property",
    "package-json-has-a-version-property",
    "package-json-has-a-license-property",
    "package-json-has-a-mit-license",
    "package-json-has-a-script-section",
    "package-json-has-no-ci-script",
    "package-json-has-no-sass-dependency-fat-useless",
    "package-json-has-no-cross-var-dependency-old-deprecated",
    "package-json-has-no-tslint-dependency-deprecated",
    "package-json-has-no-eslint-plugin-promise-5-dependency-require-eslint-7",
    "package-json-has-no-patch-precision",
    "package-json-one-unit-testing-dependency-from-vitest-mocha-uvu",
    "package-json-assert-dependency-used-in-import-equal-from-uvu-assert-instead-works-also-as-deep-equal-alternative",
    "package-json-has-no-fat-color-dependency-use-shuutils-or-nanocolors",
    "package-json-has-no-fat-fs-extra-dependency-use-native-fs",
    "package-json-has-no-utopian-shuunen-stack-dependency",
    "package-json-has-no-fat-slow-jsdom-dependency-use-happy-dom-instead",
    "package-json-has-no-fat-task-runner-use-pnpm-xyz-pnpm-abc-for-sequential-or-zero-deps-package-npm-parallel",
    "readme-md-has-a-title",
    "readme-md-has-no-link-to-deprecated-netlify-com",
    "readme-md-has-no-links-without-https-scheme",
    "readme-md-has-no-crlf-windows-carriage-return",
    "readme-md-has-no-star-flavored-list",
    "readme-md-has-no-deprecated-bettercodehub-com-badge",
    "readme-md-has-no-deprecated-david-dm-org-badge",
    "readme-md-has-no-deprecated-lgtm-com-badge",
    "readme-md-has-no-deprecated-scrutinizer-ci-com-badge",
    "readme-md-has-no-deprecated-libraries-io-badge",
    "has-a-repo-checker-json-file",
    "tsconfig-json-has-a-schema-declaration",
    "tsconfig-json-has-an-include-section",
    "tsconfig-json-has-a-verbatim-module-syntax-compiler-option",
    "tsconfig-json-has-a-out-dir-compiler-option",
    "tsconfig-json-has-a-module-resolution-compiler-option",
    "tsconfig-json-has-a-non-empty-lib-compiler-option",
    "tsconfig-json-my-folder-is-not-needed-in-include-section-my-folder-is-enough",
  ],
  "warnings": [
    "use-dependency-cruiser",
    "has-a-npmrc-file",
    "package-json-main-file-maximum-size-is-specified-in-data-file-ex-max-size-ko-100",
    "package-json-could-have-a-schema-declaration-like-schema-https-json-schemastore-org-package-43m30m-fixable-39m49m",
    "package-json-could-have-a-bugs-property-bugs",
    "package-json-could-have-a-files-property-files-n",
    "package-json-could-have-a-homepage-property-homepage",
    "package-json-could-have-a-keywords-property-keywords-n",
    "package-json-could-have-a-private-property-private-falsetrue-n",
    "package-json-could-have-a-repository-property-repository-n",
    "package-json-could-have-a-pre-script-for-version-automation-like-preversion-pnpm-check",
    "package-json-could-have-a-post-script-for-version-automation-like-postversion-git-push-git-push-tags",
    "package-json-could-have-pnpm-instead-of-npm-run-use-pnpm-instead-of-npm-run-for-performance-43m30m-fixable-39m49m",
    "package-json-could-have-a-check-script-like-check-pnpm-build-pnpm-lint",
    "package-json-could-have-a-final-echo-for-task-build-build-echo-wsbuild-w",
    "package-json-one-coverage-dependency-from-nyc-c8-v8",
    "package-json-could-have-latest-version-of-repo-checker-like-repo-check-1-38-43m30m-fixable-39m49m",
    "package-json-ts-node-without-transpile-only-detected-in-files-package-json",
    "package-json-could-have-not-fat-ts-runner-use-typescript-run-like-dev-ts-run-src-watch-or-ts-run-src-w-src-another-folder-not-fat-ts-runner-use-typescript-run-like-dev-ts-run-src-watch-or-ts-run-src-w-src-another-folder",
    "readme-md-could-have-a-logo-image-logofolder-any-file-ext",
    "readme-md-could-have-a-demo-image-or-gif-demofolder-any-file-ext",
    "readme-md-could-have-a-project-license-badge-project-licensehttps-img-shields-io-github-license-shuunen-a-great-repo-svgcolorinformationalhttps-github-com-shuunen-a-great-repo-blob-master-license-43m30m-fixable-39m49m",
    "readme-md-could-have-a-code-climate-maintainability-badge-code-climate-maintainabilityhttps-img-shields-io-codeclimate-maintainability-shuunen-a-great-repohttps-codeclimate-com-github-shuunen-a-great-repo-43m30m-fixable-39m49m",
    "readme-md-could-have-a-thanks-section-thanks",
    "readme-md-could-have-a-star-chart-stargazers-over-time-43m30m-fixable-39m49m",
    "readme-md-could-have-a-page-views-counter-free-website-counterhttps-www-websitecounterfree-com-c-phpd9id60667s12-43m30m-fixable-39m49m",
    "has-a-renovate-json-file",
  ],
}
`;

exports[`check D data folders and throw 1`] = `"failed at validating at least one rule in one folder : has-a-editorconfig-file, has-a-gitignore-file, has-a-license-file, has-a-nvmrc-file, package-json-do..."`;

exports[`check E data folders and not throw 1`] = `
{
  "failed": [
    "has-a-editorconfig-file",
    "has-a-gitignore-file",
    "has-a-license-file",
    "has-a-nvmrc-file",
    "package-json-does-not-have-a-name-property-name",
    "package-json-does-not-have-a-typescript-build-or-check-btscb",
    "package-json-does-not-have-prepare-instead-of-prepublish-deprecated-prepublish",
  ],
  "passed": [
    "avoid-main-branch-reference-use-master-instead-git-bclean",
    "has-a-package-json-file",
    "package-json-has-a-description-property",
    "package-json-has-a-author-property",
    "package-json-has-a-version-property",
    "package-json-has-a-license-property",
    "package-json-has-a-mit-license",
    "package-json-has-a-script-section",
    "package-json-has-no-ci-script",
    "package-json-has-no-cross-var-dependency-old-deprecated",
    "package-json-has-no-tslint-dependency-deprecated",
    "package-json-has-no-eslint-plugin-promise-5-dependency-require-eslint-7",
    "package-json-has-no-patch-precision",
    "package-json-one-unit-testing-dependency-from-vitest-mocha-uvu",
    "package-json-assert-dependency-used-in-import-equal-from-uvu-assert-instead-works-also-as-deep-equal-alternative",
    "package-json-has-no-fat-color-dependency-use-shuutils-or-nanocolors",
    "package-json-has-no-fat-fs-extra-dependency-use-native-fs",
    "package-json-has-no-utopian-shuunen-stack-dependency",
    "package-json-has-no-fat-slow-jsdom-dependency-use-happy-dom-instead",
    "package-json-has-no-fat-task-runner-use-pnpm-xyz-pnpm-abc-for-sequential-or-zero-deps-package-npm-parallel",
    "readme-md-has-a-title",
    "readme-md-has-no-link-to-deprecated-netlify-com",
    "readme-md-has-no-links-without-https-scheme",
    "readme-md-has-no-crlf-windows-carriage-return",
    "readme-md-has-no-star-flavored-list",
    "readme-md-has-no-deprecated-bettercodehub-com-badge",
    "readme-md-has-no-deprecated-david-dm-org-badge",
    "readme-md-has-no-deprecated-lgtm-com-badge",
    "readme-md-has-no-deprecated-scrutinizer-ci-com-badge",
    "readme-md-has-no-deprecated-libraries-io-badge",
    "has-a-repo-checker-json-file",
    "tsconfig-json-has-a-schema-declaration",
    "tsconfig-json-has-an-include-section",
    "tsconfig-json-has-a-verbatim-module-syntax-compiler-option",
    "tsconfig-json-has-a-out-dir-compiler-option",
    "tsconfig-json-has-a-module-resolution-compiler-option",
    "tsconfig-json-has-a-non-empty-lib-compiler-option",
    "tsconfig-json-my-folder-is-not-needed-in-include-section-my-folder-is-enough",
  ],
  "warnings": [
    "use-dependency-cruiser",
    "has-a-npmrc-file",
    "package-json-main-file-maximum-size-is-specified-in-data-file-ex-max-size-ko-100",
    "package-json-could-have-a-schema-declaration-like-schema-https-json-schemastore-org-package-43m30m-fixable-39m49m",
    "package-json-could-have-a-bugs-property-bugs",
    "package-json-could-have-a-files-property-files-n",
    "package-json-could-have-a-homepage-property-homepage",
    "package-json-could-have-a-keywords-property-keywords-n",
    "package-json-could-have-a-private-property-private-falsetrue-n",
    "package-json-could-have-a-repository-property-repository-n",
    "package-json-could-have-a-pre-script-for-version-automation-like-preversion-pnpm-check",
    "package-json-could-have-a-post-script-for-version-automation-like-postversion-git-push-git-push-tags",
    "package-json-could-have-pnpm-instead-of-npm-run-use-pnpm-instead-of-npm-run-for-performance-43m30m-fixable-39m49m",
    "package-json-could-have-a-check-script-like-check-pnpm-build-pnpm-lint",
    "package-json-could-have-a-final-echo-for-task-build-build-echo-wsbuild-w",
    "package-json-one-coverage-dependency-from-nyc-c8-v8",
    "package-json-could-have-latest-version-of-repo-checker-like-repo-check-1-38-43m30m-fixable-39m49m",
    "package-json-ts-node-without-transpile-only-detected-in-files-package-json",
    "package-json-could-have-not-fat-ts-runner-use-typescript-run-like-dev-ts-run-src-watch-or-ts-run-src-w-src-another-folder-not-fat-ts-runner-use-typescript-run-like-dev-ts-run-src-watch-or-ts-run-src-w-src-another-folder",
    "readme-md-could-have-a-logo-image-logofolder-any-file-ext",
    "readme-md-could-have-a-demo-image-or-gif-demofolder-any-file-ext",
    "readme-md-could-have-a-project-license-badge-project-licensehttps-img-shields-io-github-license-shuunen-a-great-repo-svgcolorinformationalhttps-github-com-shuunen-a-great-repo-blob-master-license-43m30m-fixable-39m49m",
    "readme-md-could-have-a-code-climate-maintainability-badge-code-climate-maintainabilityhttps-img-shields-io-codeclimate-maintainability-shuunen-a-great-repohttps-codeclimate-com-github-shuunen-a-great-repo-43m30m-fixable-39m49m",
    "readme-md-could-have-a-thanks-section-thanks",
    "readme-md-could-have-a-star-chart-stargazers-over-time-43m30m-fixable-39m49m",
    "readme-md-could-have-a-page-views-counter-free-website-counterhttps-www-websitecounterfree-com-c-phpd9id60667s12-43m30m-fixable-39m49m",
    "has-a-renovate-json-file",
  ],
}
`;
